### ISSUE TITLE
[RFC] Support LTX-2.5 text-to-audio-video FlowGRPO

### ISSUE THREAD

**Issue by NancyFyong** [verl-project/verl-omni#537]
## Feature request

Add support for **LTX-2.5 Full/SFT one-stage text-to-audio-video FlowGRPO** in `verl-omni`, extending the existing LTX-2.3 integration instead of introducing a second LTX architecture stack.

The initial implementation targets:

- `Lightricks/LTX-2.5-Diffusers` as the model bundle;
- `transformer_full` for the Diffusers/FSDP2 LoRA Actor;
- the pinned vLLM-Omni LTX-2.5 Full one-stage rollout path;
- synchronized video and 48 kHz stereo-audio generation;
- CLAP and ImageBind rewards;
- FlowGRPO over selected joint audio-video reverse transitions.

Status: proposal.

## Motivation

`verl-omni` already supports LTX-2.3 T2AV FlowGRPO, and the pinned vLLM-Omni runtime provides native LTX-2.5 inference. Reusing the existing LTX-2 package allows the newer model to share the Agent Loop, media transport, reward manager, trainer, and LoRA synchronization infrastructure.

LTX-2.5 is not a recipe-only rename of LTX-2.3. Correct policy training must account for:

- `transformer_full` as the Full/SFT Actor checkpoint;
- the official Full/SFT dynamic-shift schedule, evaluated at the maximum sequence anchor and stretched to terminal shift `0.1`;
- the LTX-2.5 velocity and x0-space guidance convention;
- first-causal-frame absolute-position conditioning after the transformer input projection;
- separate video/audio connector embeddings and attention masks;
- the native diffusion video decoder and its NATTEN Hub kernel;
- alignment between vLLM-Omni rollout transitions and Diffusers Actor replay.

Without a version-aware adapter, training can execute with a schedule, conditioning path, or transition density that differs from rollout, invalidating the FlowGRPO policy ratio.

Related work and non-duplication:

- Merged PR #341 provides the LTX-2.3 T2AV FlowGRPO baseline reused by this proposal.
- Open PR #365 concerns VeOmni Actor support for LTX-2.3; this proposal initially uses the existing Diffusers/FSDP2 Actor path and targets LTX-2.5.
- RFC #414 concerns OmniNFT for LTX-2.3, a direct-preference algorithm; this proposal is limited to LTX-2.5 FlowGRPO policy-gradient training.
- RFC #403 covers typed multimodal rollout I/O; this proposal reuses the existing diffusion media output contract.
- No existing LTX-2.5 issue or PR was found when this RFC was filed.

### Goals

- Keep LTX-2.3 and LTX-2.5 behind the existing `LTX2Pipeline` + `flow_grpo` registration.
- Select LTX-2.5 behavior from `model_version=2.5` without changing LTX-2.3 behavior.
- Load the Full/SFT Actor from `transformer_full`.
- Preserve token-ID-native prompt processing.
- Encode positive and negative prompts together and cache video/audio connector embeddings for Actor replay.
- Reuse vLLM-Omni's LTX-2.5 request normalization, model execution, and native diffusion decoder.
- Apply the same FlowGRPO CPS/SDE transition formula on rollout and Actor sides.
- Preserve fp32 trajectory states and transition calculations.
- Reproduce the official dynamic sigma schedule and first-causal-frame conditioning.
- Publish video plus 48 kHz audio through the existing diffusion media contract.
- Provide a runnable FSDP2 LoRA recipe, documentation, CPU contract tests, and a GPU smoke path.

### Non-goals

- Distilled checkpoints.
- Two-stage generation or latent upsampling.
- Image-to-video conditioning in the initial version.
- Different video/audio CFG scales.
- Video/audio STG, modality guidance, or guidance rescale.
- DiffusionNFT or other direct-preference algorithms.
- Reimplementing the LTX-2.5 transformer, text encoder, VAE, decoder, or vLLM-Omni phase executor in `verl-omni`.
- Treating reduced-step smoke settings as a media-quality configuration.

## Design

The proposed training flow is:

```text
text prompt
  -> LTX-2 Agent Loop tokenization
  -> vLLM-Omni LTX-2.5 Full/SFT rollout
  -> selected joint video/audio CPS transitions
  -> decoded video + 48 kHz stereo audio
  -> CLAP + ImageBind reward
  -> FlowGRPO group advantage
  -> Diffusers/FSDP2 Actor replay
  -> LoRA update
  -> rollout weight synchronization
```

### Rollout adapter

The rollout adapter remains registered as:

```python
@VllmOmniPipelineBase.register("LTX2Pipeline", algorithm="flow_grpo")
class LTX23PipelineWithLogProb(LTX2Pipeline):
    ...
```

For an LTX-2.5 checkpoint it will:

1. let vLLM-Omni resolve the Full/SFT one-stage recipe;
2. validate that the request is T2AV and uses the supported guidance subset;
3. reuse native prompt preparation, transformer execution, and decode phases;
4. replace selected transitions with the replayable FlowGRPO CPS/SDE step;
5. record packed video/audio current states, aligned next states, timesteps, and log-probabilities;
6. transport video/audio connector embeddings and masks for Actor replay;
7. remove sequence-parallel audio padding before trajectory transport;
8. publish video as the primary media stream and audio as the auxiliary stream.

The rollout payload contains:

```text
all_latents
all_next_latents
all_timesteps
all_log_probs
prompt_embeds
prompt_embeds_mask
negative_prompt_embeds
negative_prompt_embeds_mask
audio_prompt_embeds
negative_audio_prompt_embeds
video_seq_len
```

### Actor adapter

The Diffusers Actor will:

1. load `LTX2VideoTransformer3DModel` from `transformer_full`;
2. split each packed trajectory state into video and audio rows;
3. expand scalar transition times to the per-token timestep inputs expected by the transformer;
4. apply the first-causal-frame absolute-position embedding after `proj_in`;
5. predict video and audio velocities in fp32;
6. apply standard CFG in x0 space using request-level sigma values;
7. evaluate the recorded next state with the same CPS/SDE formula used by rollout.

The keyframe embedding parameter is resolved when the projection hook executes rather than captured during setup, so it remains compatible with FSDP2 parameter unsharding.

### Scheduler and guidance

LTX-2.5 Full/SFT uses fp32 sigmas with:

- linear initialization from `1` to `0`;
- dynamic-shift anchors `1024` and `4096`;
- base/max shifts `0.95` and `2.05`;
- terminal shift `0.1`;
- final sigma `0`.

The initial training contract uses one shared `guidance_scale` for video and audio. Positive and negative prompts are encoded in one text-encoder batch, and Actor replay applies the same x0-space CFG formula.

Requests using different video/audio CFG scales, STG, modality guidance, or guidance rescale fail explicitly instead of being silently approximated.

### Install

Follow [`docs/start/install.md`](https://github.com/verl-project/verl-omni/blob/main/docs/start/install.md) and install the repository-pinned vLLM-Omni revision:

```bash
uv pip install -e ".[gpu]" --torch-backend=auto
uv pip install "vllm-omni @ git+https://github.com/vllm-project/vllm-omni.git@$(cat .github/vllm_omni_pin.txt)"
uv pip install -e ".[train,dev]"
```

The rollout worker must be able to load the NATTEN Hub kernel required by the native LTX-2.5 diffusion decoder, either online or from a correctly populated local Hub-kernel cache.

The default reward recipe uses CLAP and ImageBind. ImageBind requires its package and a local `imagebind_huge.pth` checkpoint. Users must review ImageBind's CC-BY-NC-SA 4.0 license before enabling it.

### Prepare weights

Accept the gated LTX-2.x license and download the Diffusers bundle:

```bash
hf auth login
hf download Lightricks/LTX-2.5-Diffusers \
  --local-dir /path/to/LTX-2.5-Diffusers
```

The recipe loads Actor weights from:

```text
/path/to/LTX-2.5-Diffusers/transformer_full
```

The native rollout decoder also expects:

```text
/path/to/LTX-2.5-Diffusers/vae/ltx-2.5-video-vae-bf16.safetensors
```

That file comes from the gated [`Lightricks/LTX-2.5`](https://huggingface.co/Lightricks/LTX-2.5) repository and must be copied into the model bundle's `vae/` directory when it is not already present.

### Prepare data

Create one-prompt-per-line files:

```text
dataset/vid_prompt/train.txt
dataset/vid_prompt/test.txt
```

Convert them to the standard parquet format:

```bash
python3 examples/flowgrpo_trainer/ltx2/prepare_data.py \
  --input_dir ./dataset/vid_prompt \
  --output_dir ./data/ltx2_t2av \
  --val_size 128
```

The command produces `train.parquet` and `test.parquet` under the output directory.

### Launch training

```bash
MODEL_PATH=/path/to/LTX-2.5-Diffusers \
DATA_DIR=./data/ltx2_t2av \
OUTPUT_DIR=./outputs/ltx2_5_t2av_flowgrpo \
IMAGEBIND_MODEL_PATH=/path/to/imagebind_huge.pth \
NUM_GPUS=8 \
ROLLOUT_TP=8 \
TOTAL_TRAINING_STEPS=100 \
bash examples/flowgrpo_trainer/ltx2/run_ltx2_5_t2av_lora.sh
```

Default recipe settings:

| Setting | Default |
| --- | --- |
| Actor | Diffusers + FSDP2 LoRA |
| LoRA | rank 64, alpha 128 |
| Rollout | vLLM-Omni, TP=8 |
| Training resolution | 512x384 |
| Validation resolution | 960x544 |
| Frames / FPS | 121 / 24 |
| Inference steps | 30 |
| Guidance | shared video/audio CFG 3.0 |
| FlowGRPO transition | CPS, noise level 0.8 |
| SDE selection | 3 non-contiguous steps from `[0, 10)` |
| Rewards | CLAP + ImageBind weighted sum |

Extra Hydra overrides can be appended to the launcher:

```bash
bash examples/flowgrpo_trainer/ltx2/run_ltx2_5_t2av_lora.sh \
  actor_rollout_ref.rollout.pipeline.height=544 \
  actor_rollout_ref.rollout.pipeline.width=960 \
  trainer.rollout_data_save_freq=10 \
  trainer.rollout_data_max_samples=16
```

LTX-2.5 one-stage dimensions must be divisible by 32, and frame counts must follow `8k+1`.

### Outputs and resume

The launcher writes:

```text
OUTPUT_DIR/checkpoints/
OUTPUT_DIR/logs/<timestamp>/<rank>.log
OUTPUT_DIR/logs/<timestamp>/rollout_videos/
OUTPUT_DIR/logs/<timestamp>/validation_videos/
```

Checkpoint and resume behavior uses the standard trainer options:

```bash
trainer.resume_mode=auto \
trainer.save_freq=50 \
trainer.max_actor_ckpt_to_keep=2
```

## Reusable interfaces

The implementation should extend existing contracts rather than add LTX-2.5-only trainer APIs:

- `LTX2Pipeline` remains the shared architecture registration.
- Existing diffusion Agent Loop token-ID transport is reused.
- Existing `DiffusionIOSpec` carries video and audio outputs.
- Existing FlowGRPO advantage and loss implementations remain unchanged.
- Existing LTX LoRA synchronization remains the policy-update path.
- Version-specific schedule and first-frame behavior stay inside the LTX adapter package.

This keeps future Actor backends or algorithms from duplicating LTX-2.5 request and trajectory semantics.

## Validation

The contribution should include:

1. Contract tests for model-version detection, `transformer_full` validation, trajectory shapes, audio padding, and output media metadata.
2. Scheduler tests for dynamic shifting, terminal sigma behavior, and selected transition alignment.
3. Actor tests for per-token timesteps, first-causal-frame embedding, CFG batch semantics, and fp32 transition calculations.
4. Rollout/Actor numerical comparison with identical prompts, seeds, schedules, latents, and selected SDE steps.
5. LTX-2.3 regression coverage proving the shared adapter remains backward compatible.
6. A real-checkpoint GPU smoke covering rollout, CLAP/ImageBind reward, Actor update, and LoRA weight synchronization.
7. Saved video/audio outputs confirming the media transport and decoder path.

## Dependencies and risks

| Item | Risk / handling |
| --- | --- |
| Gated weights | Users must accept the LTX-2.x Community License before download |
| `transformer_full` | Fail fast when the Full/SFT recipe points at the wrong transformer subfolder |
| Native DiffVAE | Require the matching LTX-2.5 decoder artifact and NATTEN Hub kernel |
| Guidance mismatch | Reject unsupported modality-specific guidance controls |
| Actor/rollout drift | Require numerical replay comparison and initial ratio near one |
| FSDP2 hook timing | Resolve the keyframe embedding after parameter unsharding |
| Upstream API changes | Keep version-specific adaptation local and test against the pinned vLLM-Omni revision |
| Memory usage | Default to 8 GPUs, TP=8, Actor micro-batch 1, and parameter/optimizer offload |
| License compatibility | Document that model and ImageBind weights are not covered by the repository's Apache-2.0 license |

## Implementation plan

- [ ] Extend the existing LTX adapter with version-aware Full/SFT scheduling.
- [ ] Add LTX-2.5 first-causal-frame Actor conditioning.
- [ ] Add native one-stage rollout transition capture.
- [ ] Add explicit validation for unsupported guidance and pipeline modes.
- [ ] Add the `transformer_full` FSDP2 LoRA recipe.
- [ ] Add CPU schedule, trajectory, CFG, output, and Actor replay tests.
- [ ] Add real-checkpoint GPU smoke coverage and saved media verification.
- [ ] Document weight preparation, usage, resource requirements, and licenses.

## Contribution

I can provide the version-aware LTX-2.5 Actor and rollout extension, Full/SFT recipe, documentation, CPU contract tests, and GPU smoke coverage. Before opening a PR, I will review every changed line, rerun the repository checks, document the exact commands, disclose AI assistance, and repeat the duplicate-work search required by the contribution policy.
